eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -652,31 +652,38 @@ object Helpers {
       }
 
       // those are the preimages to existing received htlcs
-      val preimages = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => u.paymentPreimage }.map(r => Crypto.sha256(r) -> r).toMap
+      val preimages: Map[ByteVector32, ByteVector32] = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => u.paymentPreimage }.map(r => Crypto.sha256(r) -> r).toMap
+      val failedIncomingHtlcs: Set[Long] = commitments.localChanges.all.collect {
+        case u: UpdateFailHtlc => u.id
+        case u: UpdateFailMalformedHtlc => u.id
+      }.toSet
 
-      val htlcTxs: Map[OutPoint, Option[HtlcTx]] = localCommit.htlcTxsAndRemoteSigs.collect {
+      val htlcTxs: Map[OutPoint, LocalCommitPublished.HtlcOutputStatus] = localCommit.htlcTxsAndRemoteSigs.collect {
         case HtlcTxAndRemoteSig(txInfo@HtlcSuccessTx(_, _, paymentHash, _, _), remoteSig) =>
           if (preimages.contains(paymentHash)) {
             // incoming htlc for which we have the preimage: we can spend it immediately
             txInfo.input.outPoint -> withTxGenerationLog("htlc-success") {
               val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
               Right(Transactions.addSigs(txInfo, localSig, remoteSig, preimages(paymentHash), commitmentFormat))
-            }
+            }.map(LocalCommitPublished.HtlcOutputStatus.Spendable).getOrElse(LocalCommitPublished.HtlcOutputStatus.Unknown)
```

Comment:
I'm referring to this comment: https://github.com/ACINQ/eclair/pull/2170#discussion_r802788668
This `getOrElse` could use a new `case object TxError extends HtlcOutputStatus`

Reply:
A few arguments against this:
- the error case is not a type: for example, we don't want to pass this to serializers, right?
- we should be detected with the warning log that we are already printing
- this wouldn't be consistent with handling of errors in the generation of other transactions like _local-main-delayed_, which are even more sensitive

Reply:
Fair enough, what bothers me though is that in this case, we're actually not at all "waiting for downstream settlement", we have downstream settlement done but are erroring out...
But since this should never happen (there's no good reason transaction generation would fail) it's probably fine.

Reply:
I see your point, this `getOrElse` is clearly wrong. How about:

```scala
      val htlcTxs = localCommit.htlcTxsAndRemoteSigs.collect {
        case HtlcTxAndRemoteSig(txInfo@HtlcSuccessTx(_, _, paymentHash, _, _), remoteSig) =>
          if (preimages.contains(paymentHash)) {
            // incoming htlc for which we have the preimage: we can spend it immediately
            txInfo.input.outPoint -> withTxGenerationLog("htlc-success") {
              val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
              Right(LocalCommitPublished.HtlcOutputStatus.Spendable(Transactions.addSigs(txInfo, localSig, remoteSig, preimages(paymentHash), commitmentFormat)))
            }
          } else if (failedIncomingHtlcs.contains(txInfo.htlcId)) {
            // incoming htlc that we know for sure will never be fulfilled downstream: we can safely discard it
            txInfo.input.outPoint -> Some(LocalCommitPublished.HtlcOutputStatus.Unspendable)
          } else {
            // incoming htlc for which we don't have the preimage: we can't spend it immediately, but we may learn the
            // preimage later, otherwise it will eventually timeout and they will get their funds back
            txInfo.input.outPoint -> Some(LocalCommitPublished.HtlcOutputStatus.PendingDownstreamSettlement)
          }
        case HtlcTxAndRemoteSig(txInfo: HtlcTimeoutTx, remoteSig) =>
          // outgoing htlc: they may or may not have the preimage, the only thing to do is try to get back our funds after timeout
          txInfo.input.outPoint -> withTxGenerationLog("htlc-timeout") {
            val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
            Right(LocalCommitPublished.HtlcOutputStatus.Spendable(Transactions.addSigs(txInfo, localSig, remoteSig, commitmentFormat)))
          }
      }.collect { case (o, Some(status)) => o -> status }.toMap
```

Reply:
I think it would be worse, because our map wouldn't have an entry for every htlc output...
I'd rather keep the current version, even though the value in the `getOrElse` case is wrong, it shouldn't really happen.
If it does happen, I'd rather have an entry in the map with an invalid status than no entry in the map at all.
Does that make sense?

Reply:
> I think it would be worse, because our map wouldn't have an entry for every htlc output...

Values are now an `Option` (hence the final collect) so there would indeed be an entry for each output.

We could go on step further and make `withTxGenerationLog` return an `Either`, which we would put in our `Map`. That would be closer to your initial proposal: for each outpoint we would have either a `Success(HtlcOutputStatus)` or a `Failure`.